File: RESTFul/IMMN/Java/app1/src/main/java/com/att/api/immn/service/IMMNService.java
/* vim: set expandtab tabstop=4 shiftwidth=4 softtabstop=4 */

/*
 * ====================================================================
 * LICENSE: Licensed by AT&T under the 'Software Development Kit Tools
 * Agreement.' 2013.
 * TERMS AND CONDITIONS FOR USE, REPRODUCTION, AND DISTRIBUTIONS:
 * http://developer.att.com/sdk_agreement/
 *
 * Copyright 2013 AT&T Intellectual Property. All rights reserved.
 * For more information contact developer.support@att.com
 * ====================================================================
 */

package com.att.api.immn.service;

import java.text.ParseException;

import org.apache.commons.lang3.StringUtils;
import org.json.JSONArray;
import org.json.JSONObject;

import com.att.api.oauth.OAuthToken;
import com.att.api.rest.APIResponse;
import com.att.api.rest.RESTClient;
import com.att.api.rest.RESTException;
import com.att.api.service.APIService;

/**
 * Used to interact with version 1 of the In-app Messaging from Mobile
 * Number(IMMN) API.
 *
 * @author pk9069
 * @author kh455g
 * @version 1.0
 * @since 1.0
 * @see <a href="https://developer.att.com/docs/apis/rest/1/In-app%20Messaging%20from%20Mobile%20Number">Documentation</a>
 */
public class IMMNService extends APIService {

    /**
     * Creates an IMMNService object.
     *
     * @param fqdn fully qualified domain name to use for sending requests
     * @param token OAuth token to use for authorization
     */
    public IMMNService(String fqdn, OAuthToken token) {
        super(fqdn, token);
    }

    
    public SendResponse sendMessage(String address, String msg) throws RESTException {
        String[] addrs = {address};
        return this.sendMessage(addrs, msg);
    }

    public SendResponse sendMessage(String[] addresses, String msg) throws RESTException {
        return this.sendMessage(addresses, msg, null, false, null);
    }

    public SendResponse sendMessage(String address, String subject, 
            boolean group) throws RESTException {
        return sendMessage(address, null, subject, group);
    }

    public SendResponse sendMessage(String address, String msg, String subject, 
            boolean group) throws RESTException {
        String[] addrs = {address};
        return sendMessage(addrs, null, subject, group);
    }

    public SendResponse sendMessage(String[] addresses, String subject, 
            boolean group) throws RESTException {
        return sendMessage(addresses, null, subject, group);
    }

    public SendResponse sendMessage(String[] addresses, String msg, 
            String subject, boolean group) throws RESTException {
        return sendMessage(addresses, msg, subject, group, null);
    }

    public SendResponse sendMessage(String address, String msg, 
            String subject, boolean group, String[] attachments) throws RESTException {
        String[] addrs = {address};
        return sendMessage(addrs, msg, subject, group, attachments);
    }

    public SendResponse sendMessage(String[] addresses, String msg, 
            String subject, boolean group, String[] attachments) throws RESTException {

        final String endpoint = getFQDN() + "/myMessages/v2/messages";

        JSONObject jsonBody = new JSONObject();
        JSONObject body = new JSONObject();
        addresses = formatAddresses(addresses);

        if (msg != null)
            body.put("text", msg);

        if (subject != null)
            body.put("subject", subject);

        // group messages must specify multiple addresses
        if (addresses.length <= 1)
            group = false;
        body.put("isGroup", group);

        JSONArray jaddrs = new JSONArray();
        for (String addr : addresses)
            jaddrs.put(addr);

        body.put("addresses", jaddrs);
        jsonBody.put("messageRequest", body);

        final RESTClient rest = new RESTClient(endpoint)
            .setHeader("Accept", "application/json")
            .setHeader("Content-Type", "application/json")
            .addAuthorizationHeader(this.getToken());

        final APIResponse response = (attachments == null) 
            ? rest.httpPost(jsonBody.toString())
            : rest.httpPost(jsonBody, attachments);

        try {
            JSONObject jobj = new JSONObject(response.getResponseBody());
            return SendResponse.valueOf(jobj);
        } catch (ParseException pe) {
            throw new RESTException(pe);
        }
    }

    public MessageList getMessageList(int limit, int offset) throws RESTException {
        return getMessageList(new MessageListArgs.Builder(limit, offset).build());
    }

    public MessageList getMessageList(MessageListArgs args) throws RESTException {
        final String endpoint = getFQDN() + "/myMessages/v2/messages";

        final RESTClient client = new RESTClient(endpoint)    
            .addAuthorizationHeader(getToken())
            .setHeader("Accept", "application/json")
            .setParameter("limit", "" + args.getLimit())
            .setParameter("offset", "" + args.getOffset());

        if (args.getMessageIds() != null) {
            String msgIds = StringUtils.join(args.getMessageIds(), ",");
            client.addParameter("messageIds", msgIds);
        }

        if (args.isFavorite() != null)
            client.addParameter("isFavorite", args.isFavorite() ? "true" : "false");

        if (args.isUnread() != null)
            client.addParameter("isUnread", args.isUnread() ? "true" : "false" );

        if (args.getType() != null)
            client.addParameter("type", args.getType().getString());

        if (args.getKeyword() != null)
            client.addParameter("keyword", args.getKeyword());

        if (args.isIncoming() != null)
            client.addParameter("isIncoming", args.isIncoming() ? "true" : "false" );

        try {
            APIResponse response = client.httpGet();
            JSONObject jobj = new JSONObject(response.getResponseBody());
            return MessageList.valueOf(jobj);
        } catch (ParseException pe) {
            throw new RESTException(pe);
        }
    }

    public Message getMessage(final String msgId) throws RESTException {
        final String endpoint = getFQDN() + "/myMessages/v2/messages/" + msgId;

        final APIResponse response = new RESTClient(endpoint)
            .addAuthorizationHeader(getToken())
            .setHeader("Accept", "application/json")
            .httpGet();

        try {
            JSONObject jobj = new JSONObject(response.getResponseBody());
            return Message.valueOf(jobj.getJSONObject("message"));
        } catch (ParseException pe) {
            throw new RESTException(pe);
        }
    }

    public MessageContent getMessageContent(String msgId, String partNumber)
            throws RESTException {

        final String endpoint = getFQDN() + "/myMessages/v2/messages/" + msgId
                + "/parts/" + partNumber;

        final APIResponse response = new RESTClient(endpoint)
            .addAuthorizationHeader(getToken())
            .setHeader("Accept", "application/json")
            .httpGet();

        String ctype = response.getHeader("Content-Type");
        String clength = response.getHeader("Content-Length");
        String content = response.getResponseBody();
        return new MessageContent(ctype, clength, content);
    }

    public DeltaResponse getDelta(final String state) throws RESTException {
        final String endpoint = getFQDN() + "/myMessages/v2/delta";

        final APIResponse response = new RESTClient(endpoint)
            .addAuthorizationHeader(getToken())
            .setHeader("Accept", "application/json")
            .addParameter("state", state)
            .httpGet();

        try {
            JSONObject jobj = new JSONObject(response.getResponseBody());
            return DeltaResponse.valueOf(jobj);
        } catch (ParseException pe) {
            throw new RESTException(pe);
        }
    }

    public void updateMessages(DeltaChange[] messages) throws RESTException {
        final String endpoint = getFQDN() + "/myMessages/v2/messages";

        JSONArray jmsgs = new JSONArray();
        for (final DeltaChange change : messages) {
            JSONObject jchange = new JSONObject();
            jchange.put("messageId", change.getMessageId());
            
            if (change.isUnread() != null)
                jchange.put("isUnread", change.isUnread());

            if (change.isFavorite() != null)
                jchange.put("isFavorite", change.isFavorite());

            jmsgs.put(jchange);
        }

        JSONObject jobj = new JSONObject();
        jobj.put("messages", jmsgs);

        final APIResponse response = new RESTClient(endpoint)
            .addAuthorizationHeader(getToken())
            .setHeader("Accept", "application/json")
            .setHeader("Content-Type", "application/json")
            .httpPut(jobj.toString());

        if (response.getStatusCode() != 204) {
            final int code = response.getStatusCode();
            final String body = response.getResponseBody();
            throw new RESTException(code, body);
        }
    }

    public void updateMessage(String msgId, Boolean isUnread,
            Boolean isFavorite) throws RESTException {

        final String endpoint = getFQDN() + "/myMessages/v2/messages/" + msgId;

        JSONObject jmsg = new JSONObject();
        if (isUnread != null) jmsg.put("isUnread", isUnread);
        if (isFavorite != null) jmsg.put("isFavorite", isFavorite);
        
        JSONObject jobj = new JSONObject();
        jobj.put("message", jmsg);

        final APIResponse response = new RESTClient(endpoint)
            .addAuthorizationHeader(getToken())
            .setHeader("Accept", "application/json")
            .setHeader("Content-Type", "application/json")
            .httpPut(jobj.toString());

        if (response.getStatusCode() != 204) {
            final int code = response.getStatusCode();
            final String body = response.getResponseBody();
            throw new RESTException(code, body);
        }
    }

    public void deleteMessages(String[] msgIds) throws RESTException {
        final String endpoint = getFQDN() + "/myMessages/v2/messages";

        String msgIdsStr = StringUtils.join(msgIds, ",");

        final APIResponse response = new RESTClient(endpoint)
            .setHeader("Accept", "application/json")
            .addAuthorizationHeader(getToken())
            .addParameter("messageIds", msgIdsStr)
            .httpDelete();

        if (response.getStatusCode() != 204) {
            final int code = response.getStatusCode();
            final String body = response.getResponseBody();
            throw new RESTException(code, body);
        }
    }

    public void deleteMessage(String msgId) throws RESTException {
        final String endpoint = getFQDN() + "/myMessages/v2/messages/" + msgId;

        final APIResponse response = new RESTClient(endpoint)
            .setHeader("Accept", "application/json")
            .addAuthorizationHeader(getToken())
            .httpDelete();

        if (response.getStatusCode() != 204) {
            final int code = response.getStatusCode();
            final String body = response.getResponseBody();
            throw new RESTException(code, body);
        }
    }

    public void createMessageIndex() throws RESTException {
        final String endpoint = getFQDN() + "/myMessages/v2/messages/index";

        final APIResponse response = new RESTClient(endpoint)
            .setHeader("Accept", "application/json")
            .addAuthorizationHeader(getToken())
            .httpPost();

        if (response.getStatusCode() != 202) {
            final int code = response.getStatusCode();
            final String body = response.getResponseBody();
            throw new RESTException(code, body);
        }
    }

    public MessageIndexInfo getMessageIndexInfo() throws RESTException {
        final String endpoint = getFQDN() + "/myMessages/v2/messages/index/info";

        final APIResponse response = new RESTClient(endpoint)
            .setHeader("Accept", "application/json")
            .addAuthorizationHeader(getToken())
            .httpGet();

        try {
            JSONObject jobj = new JSONObject(response.getResponseBody());

            return MessageIndexInfo.valueOf(jobj);
        } catch (ParseException pe) {
            throw new RESTException(pe);
        }
    }

    public NotificationConnectionDetails getNotificationConnectionDetails(
            String queues) throws RESTException {
            
        final String endpoint = getFQDN()
                + "/myMessages/v2/notificationConnectionDetails";

        final APIResponse response = new RESTClient(endpoint)
            .setHeader("Accept", "application/json")
            .addAuthorizationHeader(getToken())
            .setParameter("queues", queues)
            .httpGet();

        try {
            JSONObject jobj = new JSONObject(response.getResponseBody());
            return NotificationConnectionDetails.valueOf(jobj);
        } catch (ParseException pe) {
            throw new RESTException(pe);
        }
    }

}

File: RESTFul/IMMN/PHP/app1/lib/IMMN/IMMNService.php
<?php
namespace Att\Api\IMMN;

/*
 * Copyright 2014 AT&T
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

require_once __DIR__ . '../../Srvc/APIService.php';

require_once __DIR__ . '/DeltaResponse.php';
require_once __DIR__ . '/MessageList.php';
require_once __DIR__ . '/MessageContent.php';
require_once __DIR__ . '/MessageIndexInfo.php';
require_once __DIR__ . '/NotificationCD.php';


use Att\Api\Restful\RestfulRequest;
use Att\Api\Srvc\Service;
use Att\Api\Srvc\ServiceException;
use Att\Api\Srvc\APIService;
use Att\Api\OAuth\OAuthToken;
use Att\Api\Restful\HttpGet;
use Att\Api\Restful\HttpPost;
use Att\Api\Restful\HttpPut;
use Att\Api\Restful\HttpMultipart;

use Att\Api\IMMN\DeltaResponse;
use Att\Api\IMMN\IMMNMessageList;
use Att\Api\IMMN\IMMNMessageContent;
use Att\Api\IMMN\IMMNMessageIndexInfo;
use Att\Api\IMMN\IMMNNotificactionCD;

/**
 * Used to interact with version 2 of the In-app Messaging from Mobile Number 
 * (IMMN) API.
 *
 * @category API
 * @package  IMMN
 * @author   pk9069
 * @license  http://www.apache.org/licenses/LICENSE-2.0
 * @version  Release: @package_version@ 
 * @link     https://developer.att.com/docs/apis/rest/1/In-app%20Messaging%20from%20Mobile%20Number
 */
class IMMNService extends APIService
{

    /**
     * Creates an IMMNService object that can be used to interact with
     * the IMMN API.
     *
     * @param string     $FQDN  fully qualified domain name to which requests 
     *                          will be sent
     * @param OAuthToken $token OAuth token used for authorization 
     */
    public function __construct($FQDN, OAuthToken $token) 
    {
        parent::__construct($FQDN, $token);
    }


    /**
     * Sends a message to the specified addresses. 
     *
     * @param array       $addresses strings that holds the addresses to which 
     *                               the specified messages will be sent. 
     * @param string|null $text      text body of message or null if none
     * @param string|null $subject   subject of message or null if none
     * @param array|null  $fnames    file names of attachments or null if none 
     * @param bool|null   $isGroup   whether to send as broadcast or null to
     *                               use default
     *
     * @return string message id 
     * @throws ServiceException if API request was not successful
     */
    public function sendMessage(
        $addresses, $text, $subject, $fnames = null, $isGroup = null
    ) {

        $endpoint = $this->getFqdn() . '/myMessages/v2/messages';

        $req = new RESTFulRequest($endpoint);
        $req
            ->setHeader('Accept', 'application/json')
            ->setHeader('Content-Type', 'application/json')
            ->setAuthorizationHeader($this->getToken());

        $vals = array('addresses' => $addresses);
        $vals['isGroup'] = $isGroup ? 'true' : 'false';

        if ($text !== null) { $vals['text'] = $text; }
        if ($subject !== null) { $vals['subject'] = $subject; }

        $messageRequest = array('messageRequest' => $vals);

        $jvals = json_encode($messageRequest);

        $result = null;
        if ($fnames == null) { // no attachments; send basic POST
            $req->setHeader('Content-Type', 'application/json');
            $httpPost = new HttpPost();
            $httpPost->setBody($jvals);
            $result = $req->sendHttpPost($httpPost);
        } else { // attachments; send as multipart 
            $mpart = new HttpMultipart();
            $mpart->addJSONPart($jvals); 
            foreach ($fnames as $fname) {
                $mpart->addFilePart($fname);
            }
            $result = $req->sendHttpMultipart($mpart);
        }

        $responseArr = Service::parseJson($result);
        return $responseArr['id'];
    }

    public function getMessageList($limit, $offset, $msgIds=null,
        $isUnread=null, $type=null, $keyword=null, $isIncoming=null, $isFav=null
    ) {

        $endpoint = $this->getFqdn() . '/myMessages/v2/messages';

        $req = new RestfulRequest($endpoint);

        $req
            ->setHeader('Accept', 'application/json')
            ->setAuthorizationHeader($this->getToken());

        $httpGet = new HttpGet();

        $httpGet
            ->setParam('limit', $limit)
            ->setParam('offset', $offset);
        
        if ($msgIds != null) {
            $msgIdsStr = implode(',', $msgIds);
            $httpGet->setParam('messageIds', $msgIdsStr);
        }

        if ($isUnread !== null) {
            $httpGet->setParam('isUnread', $isUnread ? 'true' : 'false');
        }
        if ($isIncoming !== null) {
            $httpGet->setParam('isIncoming', $isIncoming ? 'true' : 'false');
        }
        if ($isFav !== null) {
            $httpGet->setParam('isFavorite', $isFav ? 'true' : 'false');
        }
        
        if ($type != null) $httpGet->setParam('type', $type);
        if ($keyword != null) $httpGet->setParam('keyword', $keyword);

        $result = $req->sendHttpGet($httpGet);
        $arr = Service::parseJson($result);

        return IMMNMessageList::fromArray($arr);
    }

    public function getMessage($msgId)
    {
        $msgId = urlencode($msgId);
        $endpoint = $this->getFqdn() . '/myMessages/v2/messages/' . $msgId;

        $req = new RestfulRequest($endpoint);

        $result = $req
            ->setHeader('Accept', 'application/json')
            ->setAuthorizationHeader($this->getToken())
            ->sendHttpGet();

        $arr = Service::parseJson($result);
        $msgArr = $arr['message'];

        return IMMNMessage::fromArray($msgArr);
    }

    public function getMessageContent($msgId, $partId)
    { 
        $msgId = urlencode($msgId);
        $partId = urlencode($partId);
        $endpoint = $this->getFqdn() . '/myMessages/v2/messages/' . $msgId
            .'/parts/'. $partId;

        $req = new RestfulRequest($endpoint);

        $result = $req
            ->setHeader('Accept', 'application/json')
            ->setAuthorizationHeader($this->getToken())
            ->sendHttpGet();

        $responseCode = $result->getResponseCode();
        if ($responseCode != 200) {
            throw new ServiceException($responseCode, $result->getResponseBody());
        }

        $arr = array();
        $arr['contentType'] = $result->getHeader('content-type');
        $arr['contentLength'] = (int) ($result->getHeader('content-length'));
        $arr['content'] = $result->getResponseBody();

        return IMMNMessageContent::fromArray($arr);
    }

    public function getMessagesDelta($state)
    {  

        $endpoint = $this->getFqdn() . '/myMessages/v2/delta';

        $req = new RestfulRequest($endpoint);

        $req
            ->setHeader('Accept', 'application/json')
            ->setAuthorizationHeader($this->getToken());

        $httpGet = new HttpGet();
        $httpGet->setParam('state', $state);

        $result = $req->sendHttpGet($httpGet);
        $arr = Service::parseJson($result);

        return DeltaResponse::fromArray($arr);
    }

    public function updateMessages($immnDeltaChanges)
    {
        $endpoint = $this->getFqdn() . '/myMessages/v2/messages';

        $req = new RestfulRequest($endpoint);

        $req
            ->setHeader('Accept', 'application/json')
            ->setHeader('Content-Type', 'application/json')
            ->setAuthorizationHeader($this->getToken());
         
        $msgsArr = array();
        foreach ($immnDeltaChanges as $immnDeltaChange) {
            $msgsArr[] = $immnDeltaChange->toArray();
        }
        
        $jMsgsArr = json_encode(array('messages' => $msgsArr));

        $result = $req->sendHttpPut(new HttpPut($jMsgsArr));

        if ($result->getResponseCode() != 204) {
            $body = $result->getResponseBody();
            throw new ServiceException($result->getResponseCode(), $body);
        }
    }

    public function updateMessage($msgId, $isUnread=null, $isFavorite=null)
    {
        $msgId = urlencode($msgId);
        $endpoint = $this->getFqdn() . '/myMessages/v2/messages/' . $msgId;

        $req = new RestfulRequest($endpoint);

        $req
            ->setHeader('Accept', 'application/json')
            ->setHeader('Content-Type', 'application/json')
            ->setAuthorizationHeader($this->getToken());
         
        $vals = array();

        if ($isUnread !== null) { 
            $vals['isUnread'] = $isUnread ? 'true' : 'false';
        }

        if ($isFavorite !== null)
            $vals['isFavorite'] = $isFavorite ? 'true' : 'false';

        $jbody = json_encode(array('message' => $vals));

        $result = $req->sendHttpPut(new HttpPut($jbody));

        if ($result->getResponseCode() != 204) {
            $body = $result->getResponseBody();
            throw new ServiceException($result->getResponseCode(), $body);
        }

    }

    public function deleteMessages($msgIds)
    {
        $msgIdsStr = implode(',', $msgIds);
        $query = http_build_query(array('messageIds' => $msgIdsStr));

        $endpoint = $this->getFqdn() . '/myMessages/v2/messages?' . $query;

        $req = new RestfulRequest($endpoint);

        $req
            ->setHeader('Accept', 'application/json')
            ->setAuthorizationHeader($this->getToken());

        $result = $req->sendHttpDelete();

        if ($result->getResponseCode() != 204) {
            $body = $result->getResponseBody();
            throw new ServiceException($result->getResponseCode(), $body);
        }
    }
    
    public function deleteMessage($msgId)
    {
        $endpoint = $this->getFqdn() . '/myMessages/v2/messages/' . $msgId;

        $req = new RestfulRequest($endpoint);

        $req
            ->setHeader('Accept', 'application/json')
            ->setAuthorizationHeader($this->getToken());

        $result = $req->sendHttpDelete();

        if ($result->getResponseCode() != 204) {
            $body = $result->getResponseBody();
            throw new ServiceException($result->getResponseCode(), $body);
        }
    }

    public function createMessageIndex()
    {
        $endpoint = $this->getFqdn() . '/myMessages/v2/messages/index';

        $req = new RestfulRequest($endpoint);

        $req
            ->setHeader('Accept', 'application/json')
            ->setAuthorizationHeader($this->getToken());

        $httpPost = new HttpPost();
        $httpPost->setBody(' '); //empty body
        $result = $req->sendHttpPost($httpPost);

        if ($result->getResponseCode() != 202) {
            $body = $result->getResponseBody();
            throw new ServiceException($result->getResponseCode(), $body);
        }
    }

    public function getMessageIndexInfo()
    {
        $endpoint = $this->getFqdn() . '/myMessages/v2/messages/index/info';

        $req = new RestfulRequest($endpoint);

        $req
            ->setHeader('Accept', 'application/json')
            ->setAuthorizationHeader($this->getToken());

        $result = $req->sendHttpGet();

        $arr = Service::parseJson($result);
        return IMMNMessageIndexInfo::fromArray($arr);
    }

    public function getNotificationConnectionDetails($queues) {
        $endpoint = $this->getFqdn() 
            . '/myMessages/v2/notificationConnectionDetails';

        $req = new RestfulRequest($endpoint);

        $req
            ->setHeader('Accept', 'application/json')
            ->setAuthorizationHeader($this->getToken());
        
        $httpGet = new HttpGet();
        $httpGet->setParam('queues', $queues);

        $result = $req->sendHttpGet($httpGet);
        $arr = Service::parseJson($result);

        return IMMNNotificactionCD::fromArray($arr);
    }

}

/* vim: set expandtab tabstop=4 shiftwidth=4 softtabstop=4: */
?>
